go-controller/hybrid-overlay/pkg/windows/service/service.go

```diff
@@ -35,42 +34,41 @@ import (
 )
 
 type handler struct {
-	fromsvc chan error
+	svcName string
+	ctx     *context.Context
+	err     error
 }
 
 // InitService is the entry point for running the daemon as a Windows
 // service. It returns an indication of whether it is running as a service;
 // and an error.
-func InitService(serviceName string) error {
+func InitService(serviceName string, ctx *context.Context) error {
 	h := &handler{
-		fromsvc: make(chan error),
+		svcName: serviceName,
+		ctx:     ctx,
+		err:     nil,
 	}
 
-	var err error
 	go func() {
-		err = svc.Run(serviceName, h)
-		h.fromsvc <- err
+		h.err = svc.Run(serviceName, h)
 	}()
 
-	// Wait for the first signal from the service handler.
-	err = <-h.fromsvc
-	if err != nil {
-		return err
-	}
-	klog.Infof("Running %s as a Windows service!", serviceName)
-	return nil
+	return h.err
```

Comment:
This is always going to return `nil` now, because svc.Run() won't have completed yet by the time this function returns since it's called from a goroutine.

Reply:
thanks for pointing out, added an error channel for handling the error inside goroutine.